### DISCUSSION TITLE
Champion "range patterns"

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#8783] [Language Ideas]
- [ ] Proposal added
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

The idea is to have a pattern with a syntax like

``` antlr
range_pattern
    : 'in' constant_expression 'to' constant_expression
    ;
```

Which matches input primitive values that are `>=` the first constant expression and `<=` the second constant expression.

An alternative is to have *relational patterns*, which would be a pattern like `<3`:

``` antlr
relational_pattern
    : relational_operator constant_expression
    ;
relational_operator
    : '<'
    | '>'
    | '<='
    | '>='
    ;
```

I (@gafter) think I prefer the alternative of relational patterns, especially in conjunction with [pattern combinators like `and`](https://github.com/dotnet/csharplang/issues/1350).

**Comment by yaakov-h** [comment]
Why use a new syntax here, `a to b`, and not range syntax, `a..b`? Is this to allow for the feature to work with custom operators?

**Comment by CyrusNajmabadi** [comment]
aren't range exclusive?  (and they're a value, not a pattern).  This wants to be inclusive right?

**Comment by alrz** [comment]
If we ever get inclusive ranges, I'd expect the same syntax for range patterns.

**Comment by YairHalberstadt** [comment]
For floats, how would you match exclusively?

**Comment by CyrusNajmabadi** [comment]
Wouldn't inclusive/exclusive always be about if you're doing `<=` vs `<`?

**Comment by YairHalberstadt** [comment]
Yep, so it's useful to be able to pick.

**Comment by qrli** [comment]
The old dilemma
```cs
double v = 1.0;
switch (v) {
  case in 0 to 1.0: Good(); break;
  case in 1.0 to 2.0: NotSoGood(); break;
  default: break;
}
```

**Comment by alrz** [comment]
you'd probably get a subsumption error for that?

**Comment by jveselka** [comment]
I think there should be both `in x to y` with inclusive upper bound and `in x until y` with exclusive upper bound.

**Comment by orthoxerox** [comment]
@alrz but how would you avoid it?

**Comment by Richiban** [comment]
Is there also a way to create a System.Range with this syntax?

I really think that one of the overall goals of putting patterns into C# should be construction/deconstruction _symmetry_. 

I know that this is `case in` rather than `case`, but I still think that's it's part of the same language feature that I should be able to write:

```csharp
var r = 1 to 10;
```

Because, currently, if I want the range of numbers from 1 to 10 I will have to write

```csharp
var r = 1..11;
```

which makes me shudder.

**Comment by tpetrina** [comment]
Why not just be explicit about it:

```
switch (x)
{
 case 0 <= a < 1: // ...
 case 1 <= a: // ...
}
```

This can play nicely with custom types as well.

**Comment by alrz** [comment]
> but how would you avoid it?

@orthoxerox `1.0 + Double.Epsilon to 2.0` could work, but as far as I know, ~~F#~~ Rust doesn't allow exclusive range patterns, probably because it's not that useful in practice (and might be confusing).

**Comment by Thaina** [comment]
I don't see any difference between `..` and `to` but I want to have only 1 syntax here. If we would write `..` for a range I would like it to be that syntax for range pattern and vise versa, if we use `to` for range pattern I think we should also use `to` for range object

But then again we actually could use `when` so range pattern is not much useful? It also not ambiguous between inclusive and exclusive

```C#
case int i when i > min && i <= max:
break;
```

**Comment by YairHalberstadt** [comment]
One advantage of the range pattern is it makes it easier for the compiler to verify that all cases have been dealt with.

**Comment by YairHalberstadt** [comment]
How would you specify open ranges?

**Comment by gafter** [comment]
@Thaina The difference is that `..` is an expression that designates a half-open interval.  The `in expression to expression` form is a pattern that designates a closed interval.  Having the same syntax would imply that the latter could match the former.  Since the interval pattern is not intended to match the range expression, it is good that they do not have the same syntax.

**Comment by Neme12** [comment]
Could we have both half-open as well as closed range patterns? Maybe `in a..b` to match range expressions?

**Comment by gafter** [comment]
@Neme12 Why? Is that something people would actually use?

**Comment by gafter** [comment]
> you'd probably get a subsumption error for that?

No, since each case handles some inputs that are not handled by previous cases.

**Comment by alrz** [comment]
`1.0` overlaps between the two pattern. likewise, I'd expect an error for `1 or 2` with a previous pattern like `0 or 1`.

>each case handles some inputs that are not handled by previous cases.

I think that's true for patterns with different targets, for instance `(0, 1)` and `(0, 2)`.

Edit: It might be useful to not err on overlapping ranges for practicality reasons, but I don't think we should permit that for `or` patterns which are roughly equivalent to multi-case switch sections.

**Comment by qrli** [comment]
I'm not sure what usecases the `in` pattern intends to solve. But as I understand, the `a..b` expression, as discussed here, has at least 2 sets of usecases, where one fits `[a, b)` naturally while another is more natural to be `[a, b]`, though the former scenarios win.

So I don't see inclusive/exclusive really a differentiating factor for `..` vs `in`. I'd think both will face the 2 sets of usecases, though which one is the major one could be different.

For expression vs pattern, it may matter to you language designers, but I don't think average devs care or differentiate them. The `or` pattern may have a significant reason to not to reuse `||`, but what's the reason for `in`?

**Comment by Thaina** [comment]
@gafter I was also meant as @Neme12 that if both feature would adopt same syntax it should also mean the same range, if both use `..` it would be half open range, and if it use `to` both would be closed range

Then again we could also allow both syntax, I mean what if we adopt both syntax as syntax for `Range` then let both `switch` and `foreach` can use `in Range` as the same

```C#
void DoSomething(float f)
{
    switch(f)
    {
        case in 0 to 9:
            //match 0 - 9
        case in 0..10:
            //match 0 - 9.999....
    }

    foreach(int x in 0 to 9)
    {
        //Do 10 things
    }
}
```

**Comment by MkazemAkhgary** [comment]
:(  How about inclusive ranges be like

    in [0, 10]

Exclusive end

    in [0, 10)

Or any combination of [, ], (, ) instead of `.., to`

**Comment by Thaina** [comment]
@MkazemAkhgary I think we have discussed about this so long that we can't reuse `[]` and `()` because it very confusing and might collide with existing syntax in brace language like C#

At least `(,)` is tuple and `[,]` is indexer. So we cannot use them in combination

But maybe we just need `to` keyword

```C#
(0 to 10)
(0 to 9]
[1 to 10)
[1 to 9]
```

**Comment by thargol1** [comment]
Some 'real world' examples that might help form an opinion.

Example:
```csharp

// inclusive ranges
bool IsHexLetter(char c) => (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f'); // plain old code
bool IsHexLetter(char c) => c in ('A' to 'F') || c in ('a'..'f');
bool IsHexLetter(char c) => c in ('A' to 'F', 'a' to 'f');
bool IsHexLetter(char c) => 'A' <= c <= 'F' || 'a' <= c <= 'f';
bool IsHexLetter(char c) => c is in 'A' to 'F' or in 'a' to 'f'; // from @gafter 
bool IsHexLetter(char c) => c in 'A' to 'F' or in 'a' to 'f'; // from @alrz, same without is
bool IsHexLetter(char c) => c is char h when (h >= 'A' && h <= 'F') || (h >= 'a' && h <= 'f');

int diceRoll = Math.Random(1 to 6);
int diceRoll = Math.Random((1 to 6));

// exclusive ranges / patterns
double grade = GetGrade();
switch(grade)
{
	// pick the best option....
	case 0.0 .. 6.0: return "failed";
	case 0.0 upto 6.0: return "failed";
	case 0.0 to <6.0: return "failed"; 
	case 0.0 <= grade < 6.0: return "failed";
	case [0.0 to 6.0): return "failed";
	case is double d when 0.0 <= d && d < 6.0: return "failed";
}

// other
bool isValidProtocol(string p) => p.ToUpperInvariant() in ("HTTP", "HTTPS"); // Is string ranges a thing??
```

For me personally: I like it when code becomes more readable / intuitive. I really don't care if it's more difficult for the compiler...

Edit: 1) Added pattern from @gafter and @alrz. 2) Added more code from below.

**Comment by gafter** [comment]
With the use of the `or` pattern, this would be
``` c#
bool IsHexLetter(char c) => c is in 'A' to 'Z' or in 'a' to 'z';
```

**Comment by Neme12** [comment]
@gafter Is it intentional that you left out `is`?

**Comment by alrz** [comment]
Is there any ambiguities that `in` resolves? I think the pattern reads just fine without it.

**Comment by gafter** [comment]
If we add relational patterns (like `< 3`) then we might not need range patterns.

**Comment by YairHalberstadt** [comment]
> If we add relational patterns (like < 3) then we might not need range patterns.

Only if we also can `or` and `and` patterns

**Comment by Thaina** [comment]
@gafter I am more fond of relational pattern. Is there any disadvantage or ambiguous about it?

**Comment by qrli** [comment]
Like the `when` in `switch`?
```cs
bool IsHexLetter(char c) => c is char h when (h >= 'A' && h <= 'Z') || (h >= 'a' && h <= 'z');
```
Isn't the old way simpler?
```cs
bool IsHexLetter(char c) => (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z');
```

**Comment by thargol1** [comment]
I agree with @qrli. Whatever you decide on, it has to be more readable and preferably shorter than the current way:

``` csharp
bool IsHexLetter(char c) => (c >= 'A' && c <= 'F') || (c >= 'a' && c <= 'f'); 

switch(grade)
{
	case is double d when 0.0 <= d && d < 6.0: return "failed";
}
```

And it has to be something you can use as a parameter as well:
``` csharp
int diceRoll = Math.Random(<inclusive range parameter>);
```

**Comment by thargol1** [comment]
I was thinking... what if the `<expression1> in <expression2>` operator/command compiles to `expression2.Contains(expression1)`.
Similar like collection initializers compile to `collection.Add`.

And what if `1 to 5` compiles to `new InclusiveRange(1,5)` and the InclusiveRange-object would have a `Contains` method.

This would allow for a lot interesting stuff.

```csharp
static class OddIntegers
{
	public static bool Contains(int value) => (value & 1) == 1;
}

int value = 23;
switch(value)
{
	case in OddIntegers : return "Odd number";
}

value = 12
if (value in OddIntegers) return;
if (value in 2 to 6) return;

List<string> validPlaces = new List<string> {"New York", "Amsterdam", "Brussels"};
string place = "Eindhoven";
if (place in validPlaces) return "Yeeeeeeh";
```

**Comment by YairHalberstadt** [comment]
Currently a switch statement is usually very fast. Even with recursive pattern matching, it will be rare to find a particularly expensive switch statement/expression.

Compiling `a in b` to `b.Contains(a)` could hide some really non-performant code. Iterating through a large list can be very expensive.

Do we want to allow that?

**Comment by thargol1** [comment]
`Contains` can be slow, but consider this code for inclusive ranges:
``` csharp
public struct InclusiveRange<T> where T:struct, IComparable<T>
{
    public T MinValue { get; }
    public T MaxValue { get; }

    public InclusiveRange(T minValue, T maxValue)
    {
        MinValue = minValue;
        MaxValue = maxValue;
    }

    public bool Contains(T value) => value.CompareTo(MinValue) >= 0 && value.CompareTo(MaxValue) <= 0;
}
```
This code is very fast for the cases this issue started with: 
'in' constant_expression 'to' constant_expression

Using `Contains` is just an extra to make 'in' have even more usage. 
Besides if it can be implemented at compiler level it can be made to work with older versions of the framework.

**Comment by gafter** [comment]
This issue is not about expressions for containment in ranges.  It is about *patterns* for containment in ranges.  In a switch statement or switch expression, patterns can be much more performant.

Using relational patterns and also `or` patterns and `and` patterns, our running example could be
``` c#
bool IsHexLetter(char c) => c is >= 'A' and <= 'F' or >= 'a' and <= 'f';
```
If the precedence of `and` and `or` is not clear to you, one could use a hypothetical parenthesized pattern:
``` c#
bool IsHexLetter(char c) => c is (>= 'A' and <= 'F') or (>= 'a' and <= 'f');
```

**Comment by thargol1** [comment]
> This issue is not about expressions for containment in ranges. It is about _patterns_ for containment in ranges. In a switch statement or switch expression, patterns can be much more performant.

Is that because patterns are 'constant' and can therefore be better optimized by the compiler? Last time I created compilers way back in 1990, so forgive me my ignorance.

**Comment by gafter** [comment]
@thargol1 Yes, that and also because pattern evaluation has no specified order of operations, so the compiler is more free to select an optimal order.

**Comment by thargol1** [comment]
In this proposal @gafter suggest:

```antlr
range_pattern
    : 'in' constant_expression 'to' constant_expression
    ;
```

So `in 1 to 3` means the integers 1, 2 and 3. (Inclusive range)

~~Already build is `1 .. 3`, which means the integers 1 and 2. (Half open range.)~~

What missing is the half open range on the left and half open on both sides.

Suggestion:
```antlr
range_pattern
    : 'in' constant_expression'>'? 'to' '<'?constant_expression
    ;
```

* `in 1 to 3` means the integers 1, 2 and 3. 
* `in 1> to 3` means the integers 2 and 3. 
* `in 1 to <3` means the integers 1 and 2 ~~(is equal to `1 .. 3`)~~.
* `in 1> to <3` means the integer 2.

Of course this make more sense with floating point numbers, integers is just for easy example.

```csharp
double grade = GetGrade();
switch (grade)
{
	case in 0.0 to <5.0: return "failed";
	case in 5.0 to <6.0: return "almost made it";
	case in 6.0 to <8.0: return "you did OK";
	case in 8.0 to 10.0: return "excellent";
	default: return "Invalid grade detected".
}
```

Edit: adjusted after remark from @gafter .

**Comment by quinmars** [comment]
> `in 1> to 3` means the integers 2 and 3.

It should be `in 1< to 3` because 1 is __less__ than every element in the range.

**Comment by thargol1** [comment]
> > `in 1> to 3` means the integers 2 and 3.

> It should be `in 1< to 3` because 1 is **less** than every element in the range.

I was thinking the other way round: the range starts with the first possible number greater than 1.

**Comment by Thaina** [comment]
Is it impossible to reuse `&&` and `||` in pattern instead of keyword `and` and `or` ?

**Comment by HaloFour** [comment]
@Thaina 

> Is it impossible to reuse `&&` and `||` in pattern instead of keyword `and` and `or` ?

Reusing the operators would create ambiguities for the compiler, e.g.:

```cs
if (foo is true || false) 
```

This is discussed here: https://github.com/dotnet/csharplang/issues/1350

**Comment by gafter** [comment]
> Already build is `1 .. 3`, which means the integers 1 and 2. (Half open range.)

No, there is no such pattern syntax.  You're confusing this proposal with the existing range expression which produces an object of type `System.Range`.  It doesn't test containment.

**Comment by thargol1** [comment]
Combined with #1350 and #1338 the syntax could be something like this (it's not correct yet, the c-sharp example below does not match the syntax, I'm new to antlr):
```antlr
IN : 'in' ;
TO : 'to' ;
LPAREN : '(' ;
RPAREN : ')' ;
GT : '>' ;
LT : '<' ;
AND : 'and' ;
OR  : 'or' ;
NOT: 'not' ;
binary: AND | OR ;

range_pattern
	: IN constant_expression GT? TO LT? constant_expression
	;

pattern
	: range_pattern | is_pattern

full_pattern
	: LPAREN? NOT? pattern (binary full_pattern)* RPAREN?
```

Thus allowing for optional parentheses, binary composition with 'and' or 'or', and negation with 'not'.

Allowing for this code:
```csharp
switch (obj)
{
	case is int i and i not in 1 to 10: return "An integer not in range 1 to 10";
	case is double d and d in 5.0 to <10.0: return "Yep";
	case is char c and (c in 'A' to 'F' or c in 'a' to 'f'): return "A hex letter";
}
```

**Comment by alrz** [comment]
> Since the interval pattern is not intended to match the range expression, it is good that they do not have the same syntax.

I wouldn't miss it if a range pattern would be only compatible with some of primitive types so anything else would be a compile-time error, therefore, it won't give the impression that a range pattern should match a range expression in runtime.

**Comment by Neme12** [comment]
@thargol1 That is wrong, there is no such thing as `full_pattern` or `is_pattern`. A pattern can either be a binary pattern (and/or), a parenthesized pattern, a range pattern, a constant pattern, a declaration pattern or a recursive pattern.

Instead of
```c#
case is int i and i not in 1 to 10:
case is double d and d in 5.0 to <10.0:
case is char c and (c in 'A' to 'F' or c in 'a' to 'f'):
```
you'd have
```c#
case int i and not in 1 to 10:
case double d and in 5.0 to 10.0:
case char c and (in 'A' to 'F' or in 'a' to 'f')
```
Notice that there's no `is` (it's not part of any pattern, it's part of the is expression) and you don't repeat `i` in `and i not in` (that looks more like an expression rather than a pattern to me).

**Comment by Neme12** [comment]
Actually that is wrong (depending on what the type of the value being tested is). If it's not a double, you can't say `case double d and in 5.0 to 10.0` *I think* (because the target type of the range pattern is not a double). I'm not sure if there's any way to achieve this :confused:

**Comment by gafter** [comment]
Added the following to the original post:

> An alternative is to have *relational patterns*, which would be a pattern like `<3`:
> 
> ``` antlr
> relational_pattern
>     : relational_operator constant_expression
>     ;
> relational_operator
>     : '<'
>     | '>'
>     | '<='
>     | '>='
>     ;
> ```

**Comment by HaloFour** [comment]
So, assuming `and` patterns are also implemented, we'd be able to do something like the following?

```cs
var letterGrade = gpa switch {
    > 3.0            => "A",
    > 2.0 and <= 3.0 => "B",
    > 1.0 and <= 2.0 => "C",
    > 0.0 and <= 1.0 => "D",
    _                => "F"
};
```

> An alternative is to have relational patterns, which would be a pattern like `<3`:

I `<3` it too.

**Comment by alrz** [comment]
@HaloFour I don't think we should complicate the most common use case (inclusive range) just to support any possible combinations of operators? 

In case you want to simplify multiple `case` labels to a range pattern, you don't really want to worry about >=, <= and other operators.

I'm not against relational patterns (although I expect less usefulness vs. say user-defined patterns) but I think range patterns would be still useful when you just want a simple range match.

**Comment by Thaina** [comment]
@alrz I think simple range is not actually simple. We need to remember its half open behaviour and thus we could confuse sometimes. Relational patterns always make it obvious for everyone include newbie even it a little more verbose

**Comment by alrz** [comment]
@Thania "range patterns" as originally proposed here are inclusive, and they are that way in every language I checked. there's nothing to be confused about when you say you can rewrite `case 0: case 1: ... case 9:` to `case 0 to 9` or something.

**Comment by Thaina** [comment]
@alrz Sure but as of we discussed so long. I think now we should drop the original `range pattern` and encourage the `relational pattern` as alternative `range pattern` instead

**Comment by MkazemAkhgary** [comment]
@alrz Sorry but ranges in matlab are inclusive but in python are exclusive (half open).

**Comment by Thaina** [comment]
After I see a relational pattern. I have imagine that, what if we allow pattern as a shorthand for `Func<T,bool>`

Today we write

```C#
int[] items;
items.Where((item) => item >= 0 && item < 10);
```

Maybe we could

```C#
int[] items;
items.Where(>= 0 and < 10);
```

More than that

```C#
int?[] items;
items.Where({}); // not null pattern
SomeObject[] objs;
objs.Where({ value: >= 0 and < 10 }); // obj has field value and obj.value >= 0 && obj.value < 10
```

**Comment by Joe4evr** [comment]
@Thaina #91.

**Comment by Thaina** [comment]
@Joe4evr That's right. But only allowing pattern as `Func<T,bool>` even it was more limited but it make this shorter and also covered just the most common use case

```C#
items.Where(@ >= 0 && @ < 10);
// vs
items.Where(>= 0 and < 10);
```

**Comment by HaloFour** [comment]
@Thaina 

It's a single character difference, and I disagree that it's the most common use case.  I posit that LINQ/EF against POCOs are the most common use case.

**Comment by ghost** [comment]
I wrote This in #2268 
While I am reading this example in Docs:
```C#
static string Quadrant(Point p) => p switch
{
    (0, 0) => "origin",
    (var x, var y) when x > 0 && y > 0 => "Quadrant 1",
    (var x, var y) when x < 0 && y > 0 => "Quadrant 2",
    (var x, var y) when x < 0 && y < 0 => "Quadrant 3",
    (var x, var y) when x > 0 && y < 0 => "Quadrant 4",
    (var x, var y) => "on a border",
    _ => "unknown"
};
```

I wondered if it can be shorter and more readable to achive the purpose of pattern matching, so  I suggest this:
```C#
static string Quadrant(Point p) => p switch
{
    (0, 0) => "origin",
    (> 0, > 0) => "Quadrant 1",
    (< 0, > 0) => "Quadrant 2",
    (< 0, < 0) => "Quadrant 3",
    (> 0, < 0) => "Quadrant 4",
    (_, _) => "on a border",
    _ => "unknown"
};
```

@HaloFour siad this proposal would achieve this, so I am a fan :)

**Comment by munael** [comment]
This is a special case of 'Active Patterns' (which we don't have and I can't find a proposal for them).

I'd rather have active patterns in the language, and some helpers in the library.

```cs
static string Quadrant(Point p) => p switch
{
    (0, 0) => "origin",
    (GT(0), GT(0)) => "Quadrant 1",
    (LT(0), GT(0)) => "Quadrant 2",
    (< 0, < 0) => "Quadrant 3", // With custom operators (orthogonal feature) [1]
    (GT(var x, 0), LT(var y, 0)) => $"Quadrant 4: x: {x}, y: {y}", // `GT` et al. are (overloadable) patterns.
    // the `pat(var x)` form lets the pattern expose a value/ref through a new name
    // while other expressions are just passed as if arguments to a function.
    (_, _) => "on a border",
    _ => "unknown"
};
```

If we ever get custom operators, we get exactly the proposed syntax ([1]) for free. The features are then separate.

**Comment by jho1965us** [comment]
FWIW in asn.1 you can specify both inclusive, exclusive and half open ranges the syntax seem to work pretty good.
inlusive ranges this way:

    min..max

exclusive this way

    min<..<max

half open excluding max

    min..<max

half open excluding min

    min<..max

**Comment by TheFanatr** [comment]
To keep the dots in order to make range-related syntax consistent, I'd propose the following options.

### Change Current Meaning and Extend With Colons
> Channel C#'s Inner Python

`1..3` = `(1, 3)`
`:1..3:` = `[1, 3]`
`:1..3` = `[1, 3)`
`1..3:` = `(1, 3]`

### Change Current Meaning but Keep Dots

`1..3` = `(1, 3)`
`.1..3.` = `[1, 3]`
`.1..3` = `[1, 3)`
`1..3.` = `(1, 3]`

### Keep Current Meaning and Extend With Dots

`1..3` = `[1, 3)`
`1..3.` = `[1, 3]`
`1...3` = `(1, 3)`
`1...3.` = `(1, 3]`

### Keep Current Meaning and Extend With Colons
> Channel C#'s Inner Python but More Wack

`1..3` = `[1, 3)`
`1..3:` = `[1, 3]`
`1:..3` = `(1, 3)`
`1:..3:` = `(1, 3]`

The formal mathematical equivalencies are displayed after the equal signs. The colons or extra dots essentially act as range endpoint augmenters, pulling the range endpoint in the metaphorical left or right direction for more strict or less strict on the left and less strict or more strict.

##### Example with First Option
###### Default
Syntax: `0..10`
Expression: <code>[ **(** **)** ]</code>
Formal Notation: `(0, 10)`
###### Outer Left Pulled Left
Syntax: `:0..10`
Expression: <code>**[** ( **)** ]</code>
Formal Notation: `[0, 10)`

**Comment by gafter** [comment]
I am withdrawing my championship of this issue in favor of #812
